sky/provision/kubernetes/network.py

```diff
@@ -0,0 +1,205 @@
+        network_utils.KubernetesPortMode.LOADBALANCER.value)
+    try:
+        port_mode = network_utils.KubernetesPortMode.from_str(mode_str)
+    except ValueError as e:
+        with ux_utils.print_exception_no_traceback():
+            raise ValueError(str(e) + ' Please check: ~/.sky/config.yaml.') \
+                from None
+
+    return port_mode
+
+
+def open_ports(
+    cluster_name_on_cloud: str,
+    ports: List[str],
+    provider_config: Optional[Dict[str, Any]] = None,
+) -> None:
+    """See sky/provision/__init__.py"""
+    port_mode = _get_port_mode()
+
+    if port_mode == network_utils.KubernetesPortMode.LOADBALANCER:
+        _open_ports_using_loadbalancer(
+            cluster_name_on_cloud=cluster_name_on_cloud,
+            ports=ports,
+            provider_config=provider_config)
+    elif port_mode == network_utils.KubernetesPortMode.INGRESS:
+        _open_ports_using_ingress(cluster_name_on_cloud=cluster_name_on_cloud,
+                                  ports=ports,
+                                  provider_config=provider_config)
+
+
+def _open_ports_using_loadbalancer(
+    cluster_name_on_cloud: str,
+    ports: List[str],
+    provider_config: Optional[Dict[str, Any]] = None,
+) -> None:
+    assert provider_config is not None, 'provider_config is required'
+    service_name = _LOADBALANCER_SERVICE_NAME.format(
+        cluster_name_on_cloud=cluster_name_on_cloud)
+    content = network_utils.fill_loadbalancer_template(
+        namespace=provider_config['namespace'],
+        service_name=service_name,
+        ports=ports,
+        selector_key='skypilot-cluster',
+        selector_value=cluster_name_on_cloud,
+    )
+    network_utils.create_or_replace_namespaced_loadbalancer(
+        namespace=provider_config['namespace'],
+        service_name=service_name,
+        service_spec=content['service_spec'])
+
+
+def _open_ports_using_ingress(
+    cluster_name_on_cloud: str,
+    ports: List[str],
+    provider_config: Optional[Dict[str, Any]] = None,
+) -> None:
+    if not network_utils.ingress_controller_exists():
+        raise Exception(
+            "Ingress controller not found. Please install ingress controller first."
+        )
+
+    assert provider_config is not None, 'provider_config is required'
+    for port in ports:
+        service_name = f"{cluster_name_on_cloud}-skypilot-service--{port}"
+        ingress_name = f"{cluster_name_on_cloud}-skypilot-ingress--{port}"
+        path_prefix = _PATH_PREFIX.format(
+            cluster_name_on_cloud=cluster_name_on_cloud, port=port)
+
+        content = network_utils.fill_ingress_template(
+            namespace=provider_config['namespace'],
+            path_prefix=path_prefix,
+            service_name=service_name,
+            service_port=port,
+            ingress_name=ingress_name,
+            selector_key='skypilot-cluster',
+            selector_value=cluster_name_on_cloud,
+        )
+        network_utils.create_namespaced_service(
```

Comment:
Current ports implementation is incremental, i.e. if I do `sky launch -c mycluster --ports 8080` and then `sky launch -c mycluster --ports 8081`, the second `sky launch` will update the ports to `'8080-8081'`. In https://github.com/skypilot-org/skypilot/blob/53ba415bb2f6ee72dcbe23edee765bcf6f1e1fe3/sky/backends/cloud_vm_ray_backend.py#L3009-L3017
We passed the updated list to the `open_ports` API (`'8080-8081'` in the example above). Therefore, the `open_ports` API should update the ports to what `ports` describe. I'm not very familiar with k8s API, but the API is named `create_xxx` rather than `create_or_update_xxx`; so just want to make sure, does this API support update semantic?

Reply:
Great point, I will update the ingress implementation to be `create_or_update`. LoadBalancer is already `create_or_update`.

Reply:
Updated in https://github.com/skypilot-org/skypilot/pull/2713/commits/1d95434508c678bdc461eac4ace36d85f9f0af01